src/buffer.cc

```diff
@@ -204,6 +205,82 @@ Result Buffer::CompareRMSE(Buffer* buffer, float tolerance) const {
   return {};
 }
 
+std::vector<double> Buffer::GetHistogramForChannel(uint32_t channel,
+                                                   uint32_t num_bins) const {
+  std::vector<double> bins(num_bins, 0.0);
+  auto* buf_ptr = GetValues<uint8_t>();
+  const auto& segments = format_->GetSegments();
+  const auto comp = segments[channel].GetComponent();
+  auto num_channels = format_->ValuesPerElement();
+  buf_ptr += channel * comp->SizeInBytes();
+  for (size_t i = 0; i < ElementCount(); ++i) {
+    const auto bin = *reinterpret_cast<const uint8_t*>(buf_ptr);
+    bins[bin]++;
+    // Assumes unpacked format where all channels have the same bit size
+    buf_ptr += comp->SizeInBytes() * num_channels;
+  }
+
+  return bins;
+}
+
+Result Buffer::CompareHistogramEMD(Buffer* buffer, float tolerance) const {
+  auto result = CheckCompability(buffer);
+  if (!result.IsSuccess())
+    return result;
+
+  const int num_bins = 256;
+  auto num_channels = format_->ValuesPerElement();
+
+  for (auto segment : format_->GetSegments()) {
```

Comment:
This `for` can turn into:

```
if (!format_->IsUint8() || num_channels != 4)
  return Result("EMD comparision only supports 8 bit unorm  with four channels";
```

Reply:
format_IsUint8 returns false because of type_->IsNumber check. The format type seems to be a list.

Reply:
Sigh, that's a bug in the Format::IsUint8, it should still return true in that case. Can you file a bug?

Reply:
Sure